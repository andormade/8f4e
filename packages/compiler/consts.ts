export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
