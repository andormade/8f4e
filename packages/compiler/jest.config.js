module.exports = {
	transform: {
		"^.+\\.(t|j)sx?$": ["@swc/jest", {
			jsc: {
				target: 'es2021',
			},
			sourceMaps: true,
		}],
	},
	testMatch: ['/**/*.test.ts'],
};
