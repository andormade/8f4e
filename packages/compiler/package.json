{
	"name": "@8f4e/compiler",
	"version": "1.0.0",
	"description": "",
	"main": "./dist/index.js",
	"types": "./dist/index.d.ts",
	"scripts": {
		"test": "node --experimental-vm-modules --experimental-wasi-unstable-preview1 ../../node_modules/.bin/jest",
		"build": "tsc"
	},
	"repository": {
		"type": "git",
		"url": "git+https://github.com/andormade/8f4e.git"
	},
	"author": "Andor Polgar",
	"license": "MIT",
	"bugs": {
		"url": "https://github.com/andormade/8f4e/issues"
	},
	"homepage": "https://github.com/andormade/8f4e#readme",
	"dependencies": {
		"bytecode-utils": "file:packages/bytecode-utils",
		"module-compiler": "file:packages/module-compiler"
	}
}