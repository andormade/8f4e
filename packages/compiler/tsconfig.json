{
	"compilerOptions": {
		"lib": ["esnext", "dom"],
		"module": "commonjs",
		"downlevelIteration": true,
		"moduleResolution": "node",
		"esModuleInterop": true,
		"outDir": "dist/",
		"declaration": true,
		"resolveJsonModule": true,
		"strictNullChecks": true
	},
	"include": ["src/**/*.ts", "src/**/*.asm.ts"],
	"exclude": ["dist", "node_modules"]
}
