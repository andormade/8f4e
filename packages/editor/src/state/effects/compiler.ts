import { State } from '../types';
import { EventDispatcher } from '../../events';

export default async function compiler(state: State, events: EventDispatcher) {
	const workerUrl = new URL('../../../../../packages/worker/src/index.ts', import.meta.url);

	const worker = new Worker(workerUrl, {
		type: 'module',
	});

	function onMidiMessage(data) {
		worker.postMessage({
			type: 'midimessage',
			payload: data,
		});
	}

	async function onRecompile() {
		if (!state.compiler.memoryRef) {
			return;
		}

		worker.postMessage({
			type: 'recompile',
			payload: {
				memoryRef: state.compiler.memoryRef,
				modules: Array.from(state.graphicHelper.modules).map(module => {
					return { code: module.code };
				}),
				compilerOptions: {
					...state.compiler.compilerOptions,
					constants: {
						...state.compiler.compilerOptions.constants,
						SAMPLE_RATE: { value: state.project.sampleRate, isInteger: true },
						AUDIO_BUFFER_SIZE: { value: 128, isInteger: true },
						LEFT_CHANNEL: { value: 0, isInteger: true },
						RIGHT_CHANNEL: { value: 1, isInteger: true },
					},
				},
			},
		});
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'RNBOMessage':
				events.dispatch('RNBOMessage', data.payload);
				break;
			case 'buildOk':
				state.compiler.compiledModules = data.payload.compiledModules;
				state.compiler.codeBuffer = data.payload.codeBuffer;
				state.compiler.memoryBuffer = new Int32Array(state.compiler.memoryRef.buffer);
				state.compiler.memoryBufferFloat = new Float32Array(state.compiler.memoryRef.buffer);
				state.compiler.isCompiling = false;
				state.compiler.compilationTime = (performance.now() - state.compiler.lastCompilationStart).toFixed(2);

				state.compiler.buildErrors = [];
				events.dispatch('buildOk');
				break;
			case 'buildError':
				state.compiler.buildErrors = [
					{
						lineNumber: data.payload.line.lineNumber,
						moduleId: data.payload.context.namespace.moduleName,
						code: data.payload.errorCodadde,
						message: data.payload.message,
					},
				];
				events.dispatch('buildError');
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', onRecompile);
	events.on('moduleAdded', onRecompile);
	events.on('deleteModule', onRecompile);
	events.on('init', onRecompile);
	events.on('codeChange', onRecompile);
	events.on('midiMessage', onMidiMessage);
}
