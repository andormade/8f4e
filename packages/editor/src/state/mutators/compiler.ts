import { State } from '../types';

export function compilationDone(state: State, data, memoryRef: WebAssembly.Memory): void {
	state.compiler.memoryBuffer = new Int32Array(memoryRef.buffer);
	state.compiler.memoryRef = memoryRef;
	state.compiler.memoryAddressLookup = data.payload.memoryAddressLookup;
	state.compiler.isCompiling = false;
	state.compiler.compilationTime = (performance.now() - state.compiler.lastCompilationStart).toFixed(2);
	state.compiler.compiledModules = data.payload.compiledModules;
}

export function recompile(state: State): void {
	state.compiler.isCompiling = true;
	state.compiler.lastCompilationStart = performance.now();
}
