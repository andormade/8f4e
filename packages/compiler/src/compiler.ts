import { createFunctionBody, createLocalDeclaration } from './wasmUtils/sectionHelpers';
import Type from './wasmUtils/type';
import instructions from './instructions';
import { AST, Argument, ArgumentType, CompiledModule, Namespace, MemoryMap } from './types';
import { WORD_LENGTH } from './consts';

export { MemoryTypes, MemoryMap } from './types';

function parseArgument(argument: string): Argument {
	return /^-?[0-9]+$/.test(argument)
		? { value: parseInt(argument, 10), type: ArgumentType.LITERAL }
		: { value: argument, type: ArgumentType.IDENTIFIER };
}

function parseLine(line: string, lineNumber: number): AST[number] {
	const [, instruction, ...args] = line.match(/\s*(\S+)\s*(\S*)\s*(\S*)\s*(\S*)/) || [];

	return {
		lineNumber,
		instruction,
		arguments: args
			.filter(argument => {
				return argument !== '';
			})
			.map(parseArgument),
	};
}

function isComment(line: string): boolean {
	return /\s*#/.test(line);
}

function isValidInstruction(line: string): boolean {
	return /\s*(\S+)\s*(\S*)\s*(\S*)/.test(line);
}

export function compileToAST(module: string[]) {
	return module
		.map((line, index) => [index + 1, line] as [number, string])
		.filter(([, line]) => !isComment(line))
		.filter(([, line]) => isValidInstruction(line))
		.map(([lineNumber, line]) => {
			return parseLine(line, lineNumber);
		});
}

export function compileLine(
	line: AST[number],
	namespace: Namespace,
	startingByteAddress: number
): { byteCode: number[]; namespace: Namespace } {
	if (!instructions[line.instruction]) {
		throw `1001: Unrecognized instruction: '${line.instruction}'`;
	}
	return instructions[line.instruction](line, namespace, startingByteAddress);
}

function getModuleId(ast: AST): string | undefined {
	return (
		ast
			.find(
				line => line.instruction === 'module' && line.arguments[0] && line.arguments[0].type === ArgumentType.IDENTIFIER
			)
			?.arguments[0].value.toString() || undefined
	);
}

export function compile(module: string[], startingByteAddress = 0): CompiledModule {
	const ast = compileToAST(module);
	const moduleId = getModuleId(ast);

	if (!moduleId) {
		throw '1007: Missing module ID';
	}

	let memory: MemoryMap = new Map();
	let locals: string[] = [];
	let consts: Record<string, number> = {};

	const wa = ast
		.reduce((acc, line) => {
			const { byteCode, namespace } = compileLine(line, { locals, memory, consts }, startingByteAddress);
			consts = namespace.consts;
			locals = namespace.locals;
			memory = namespace.memory;
			acc.push(byteCode);
			return acc;
		}, [] as number[][])
		.flat();

	const [, lastMemoryItem = { relativeWordAddress: 0, wordSize: 0 }] = Array.from(memory).pop() || [];

	return {
		moduleId,
		functionBody: createFunctionBody([createLocalDeclaration(Type.I32, locals.length)], wa),
		byteAddress: startingByteAddress,
		wordAddress: startingByteAddress / WORD_LENGTH,
		memoryMap: memory,
		memoryWordSize: lastMemoryItem.relativeWordAddress + lastMemoryItem.wordSize,
		ast,
	};
}
