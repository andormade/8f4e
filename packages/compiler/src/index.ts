import {
	createCodeSection,
	createExportSection,
	createFunctionBody,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import Type from './wasmUtils/type';
import { call, f32store, i32store } from './wasmUtils/instructionHelpers';
import { compile as compileModule, compileToAST } from './compiler';
import {
	AST,
	ArgumentLiteral,
	ArgumentType,
	CompileOptions,
	CompiledModule,
	CompiledModuleLookup,
	Module,
	Namespace,
	Namespaces,
} from './types';
import {
	GLOBAL_ALIGNMENT_BOUNDARY,
	I16_SIGNED_LARGEST_NUMBER,
	I16_SIGNED_SMALLEST_NUMBER,
	I32_SIGNED_LARGEST_NUMBER,
} from './consts';
import { ErrorCode, getError } from './errors';
import { sortModules } from './gaphOptimizer';
import { WASM_MEMORY_PAGE_SIZE } from './wasmUtils/consts';

export * from './types';
export { I16_SIGNED_LARGEST_NUMBER, I16_SIGNED_SMALLEST_NUMBER } from './consts';
export * as examples from './examples';
export { Instruction } from './instructions';
export { default as instructions } from './instructions';
export { instructionParser } from './compiler';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

function collectConstants(ast: AST): Namespace['consts'] {
	return Object.fromEntries(
		ast
			.filter(({ instruction }) => instruction === 'const')
			.map(({ arguments: _arguments }) => {
				return [
					_arguments[0].value,
					{
						value: parseFloat(_arguments[1].value.toString()),
						isInteger: (_arguments[1] as ArgumentLiteral).isInteger,
					},
				];
			})
	);
}

function resolveInterModularConnections(compiledModules: CompiledModuleLookup) {
	compiledModules.forEach(({ ast, memoryMap }) => {
		ast.forEach(line => {
			const { instruction, arguments: _arguments } = line;
			if (
				['int*', 'int**', 'float*', 'float**', 'init', 'int'].includes(instruction) &&
				_arguments[0] &&
				_arguments[1] &&
				_arguments[0].type === ArgumentType.IDENTIFIER &&
				_arguments[1].type === ArgumentType.IDENTIFIER &&
				/&(\S+)\.(\S+)/.test(_arguments[1].value)
			) {
				// Remove &
				const [targetModuleId, targetMemoryId] = _arguments[1].value.substring(1).split('.');

				const targetModule = compiledModules.get(targetModuleId);

				if (!targetModule) {
					throw getError(ErrorCode.UNDECLARED_IDENTIFIER, line);
				}

				const targetMemory = targetModule.memoryMap.get(targetMemoryId);

				if (!targetMemory) {
					throw getError(ErrorCode.UNDECLARED_IDENTIFIER, line);
				}

				const memory = memoryMap.get(_arguments[0].value);

				if (memory) {
					memory.default = targetMemory.byteAddress;
				}
			}
		});
	});
}

export function compileModules(modules: AST[], options: CompileOptions): CompiledModule[] {
	let memoryAddress = options.startingMemoryWordAddress;
	const builtInConsts: Namespace['consts'] = {
		I16_SIGNED_LARGEST_NUMBER: { value: I16_SIGNED_LARGEST_NUMBER, isInteger: true },
		I16_SIGNED_SMALLEST_NUMBER: { value: I16_SIGNED_SMALLEST_NUMBER, isInteger: true },
		I32_SIGNED_LARGEST_NUMBER: { value: I32_SIGNED_LARGEST_NUMBER, isInteger: true },
		WORD_SIZE: { value: GLOBAL_ALIGNMENT_BOUNDARY, isInteger: true },
		...options.environmentExtensions.constants,
	};

	const namespaces: Namespaces = new Map(
		modules.map(ast => {
			const moduleName = getModuleName(ast);
			return [moduleName, { consts: collectConstants(ast) }];
		})
	);

	return modules.map((ast, index) => {
		const module = compileModule(
			ast,
			builtInConsts,
			namespaces,
			memoryAddress * GLOBAL_ALIGNMENT_BOUNDARY,
			options.maxMemorySize,
			index
		);
		memoryAddress += module.wordAlignedSize;

		if (options.maxMemorySize * WASM_MEMORY_PAGE_SIZE <= memoryAddress) {
			throw 'Memory limit exceeded';
		}

		return module;
	});
}

export function getModuleName(ast: AST) {
	const moduleInstruction = ast.find(line => {
		return line.instruction === 'module';
	});

	if (!moduleInstruction) {
		throw 'Missing module instruction';
	}

	const argument = moduleInstruction.arguments[0];

	if (argument.type !== ArgumentType.IDENTIFIER) {
		throw 'Module instruction argument type invalid';
	}

	return argument.value;
}

export function generateMemoryInitiatorFunction(compiledModules: CompiledModule[]) {
	return compiledModules.flatMap(module => {
		let pointer = module.byteAddress;
		const instructions: number[] = [];

		Array.from(module.memoryMap.values()).forEach(memory => {
			// TODO: figure out something efficient to initialise buffers larger than 32.
			if (memory.wordAlignedSize > 1 && memory.default instanceof Map) {
				memory.default.forEach((value, relativeWordAddress) => {
					instructions.push(
						...(memory.isInteger
							? i32store(pointer + (relativeWordAddress + 1) * GLOBAL_ALIGNMENT_BOUNDARY, value)
							: f32store(pointer + (relativeWordAddress + 1) * GLOBAL_ALIGNMENT_BOUNDARY, value))
					);
				});
				pointer += memory.wordAlignedSize * GLOBAL_ALIGNMENT_BOUNDARY;
			} else if (memory.wordAlignedSize === 1 && memory.default !== 0) {
				instructions.push(
					...(memory.isInteger
						? i32store(pointer, memory.default as number)
						: f32store(pointer, memory.default as number))
				);
				pointer += GLOBAL_ALIGNMENT_BOUNDARY;
			} else if (memory.wordAlignedSize === 1 && memory.default === 0) {
				pointer += GLOBAL_ALIGNMENT_BOUNDARY;
			}
		});

		return instructions;
	});
}

export default function compile(
	modules: Module[],
	options: CompileOptions
): {
	codeBuffer: Uint8Array;
	compiledModules: CompiledModuleLookup;
	allocatedMemorySize: number;
} {
	const astModules = modules.map(({ code }) => compileToAST(code, options));
	const sortedModules = sortModules(astModules);
	const compiledModules = compileModules(sortedModules, { ...options, startingMemoryWordAddress: 1 });
	const compiledModulesMap = new Map(compiledModules.map(({ id, ...rest }) => [id, { id, ...rest }]));
	resolveInterModularConnections(compiledModulesMap);
	const functionBodies = compiledModules.map(({ functionBody }) => functionBody);
	const functionSignatures = compiledModules.map(() => 0x00);
	const cycleFunction = compiledModules.map((module, index) => call(index + 3)).flat();
	const memoryInitiatorFunction = generateMemoryInitiatorFunction(compiledModules);

	return {
		codeBuffer: Uint8Array.from([
			...HEADER,
			...VERSION,
			...createTypeSection([
				createFunctionType([], []),
				createFunctionType([Type.I32], [Type.I32]),
				createFunctionType([Type.I32, Type.I32], [Type.I32]),
			]),
			...createImportSection([
				createMemoryImport('js', 'memory', options.initialMemorySize, options.maxMemorySize, true),
			]),
			...createFunctionSection([0x00, 0x00, 0x00, ...functionSignatures]),
			...createExportSection([
				createFunctionExport('init', 0x00),
				createFunctionExport('cycle', 0x01),
				createFunctionExport('buffer', 0x02),
			]),
			...createCodeSection([
				createFunctionBody([], memoryInitiatorFunction),
				createFunctionBody([], cycleFunction),
				createFunctionBody([], new Array(128).fill(call(1)).flat()),
				...functionBodies,
			]),
		]),
		compiledModules: compiledModulesMap,
		allocatedMemorySize:
			compiledModules[compiledModules.length - 1].byteAddress +
			compiledModules[compiledModules.length - 1].wordAlignedSize * GLOBAL_ALIGNMENT_BOUNDARY,
	};
}
