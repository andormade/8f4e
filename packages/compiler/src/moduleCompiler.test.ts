import { compile, compileToAST } from './compiler';

const fixture = `
module abs

; memory
int DEFAULT_VALUE -1
int* in:1 &DEFAULT_VALUE
int out 0
int[] arr 32 -1

; registers
local int input

; code
push out
push in:1
load
load
localSet input
localGet input
push 0
lessThan
if 
    push 0
    localGet input
    sub
else
    localGet input
	ifEnd
store
moduleEnd`.split('\n');

describe('moduleCompiler', () => {
	const ast = compileToAST(fixture);

	test('ast', () => {
		expect(ast).toMatchSnapshot();
	});

	test('compiled code', () => {
		expect(compile(ast, {}, 0).functionBody).toMatchSnapshot();
	});
});
