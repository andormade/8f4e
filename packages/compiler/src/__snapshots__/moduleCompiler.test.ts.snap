// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moduleCompiler ast 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "abs",
      },
    ],
    "instruction": "module",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "DEFAULT_VALUE",
      },
      Object {
        "type": "literal",
        "value": -1,
      },
    ],
    "instruction": "private",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in:1",
      },
      Object {
        "type": "identifier",
        "value": "&DEFAULT_VALUE",
      },
    ],
    "instruction": "inputPointer",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "output",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "arr",
      },
      Object {
        "type": "literal",
        "value": 32,
      },
      Object {
        "type": "literal",
        "value": -1,
      },
    ],
    "instruction": "array",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "local",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
    ],
    "instruction": "push",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in:1",
      },
    ],
    "instruction": "push",
    "lineNumber": 15,
  },
  Object {
    "arguments": Array [],
    "instruction": "load",
    "lineNumber": 16,
  },
  Object {
    "arguments": Array [],
    "instruction": "load",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 18,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "localGet",
    "lineNumber": 19,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [],
    "instruction": "lessThan",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [],
    "instruction": "if",
    "lineNumber": 22,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "localGet",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [],
    "instruction": "sub",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 26,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "localGet",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 29,
  },
]
`;

exports[`moduleCompiler compiled code 1`] = `
Array [
  44,
  1,
  1,
  127,
  65,
  8,
  40,
  2,
  0,
  65,
  4,
  40,
  2,
  0,
  40,
  2,
  0,
  40,
  2,
  0,
  40,
  2,
  0,
  33,
  0,
  32,
  0,
  65,
  0,
  72,
  4,
  127,
  65,
  0,
  32,
  0,
  107,
  5,
  32,
  0,
  11,
  54,
  2,
  0,
  11,
]
`;
